File: ios/Classes/FlutterFfmpegPlugin.m
/*
 * Copyright (c) 2019 Taner Sener
 *
 * This file is part of FlutterFFmpeg.
 *
 * FlutterFFmpeg is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FlutterFFmpeg is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with FlutterFFmpeg.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "FlutterFFmpegPlugin.h"

#import <mobileffmpeg/ArchDetect.h>
#import <mobileffmpeg/MobileFFmpegConfig.h>
#import <mobileffmpeg/MobileFFmpeg.h>
#import <mobileffmpeg/MobileFFprobe.h>

static NSString *const PLATFORM_NAME = @"ios";

static NSString *const KEY_VERSION = @"version";
static NSString *const KEY_RC = @"rc";
static NSString *const KEY_PLATFORM = @"platform";
static NSString *const KEY_PACKAGE_NAME = @"packageName";
static NSString *const KEY_LAST_RC = @"lastRc";
static NSString *const KEY_LAST_COMMAND_OUTPUT = @"lastCommandOutput";
static NSString *const KEY_PIPE = @"pipe";

static NSString *const KEY_LOG_TEXT = @"log";
static NSString *const KEY_LOG_LEVEL = @"level";

static NSString *const KEY_STAT_TIME = @"time";
static NSString *const KEY_STAT_SIZE = @"size";
static NSString *const KEY_STAT_BITRATE = @"bitrate";
static NSString *const KEY_STAT_SPEED = @"speed";
static NSString *const KEY_STAT_VIDEO_FRAME_NUMBER = @"videoFrameNumber";
static NSString *const KEY_STAT_VIDEO_QUALITY = @"videoQuality";
static NSString *const KEY_STAT_VIDEO_FPS = @"videoFps";

static NSString *const EVENT_LOG = @"FlutterFFmpegLogCallback";
static NSString *const EVENT_STAT = @"FlutterFFmpegStatisticsCallback";

/**
 * Flutter FFmpeg Plugin
 */
@implementation FlutterFFmpegPlugin {
    FlutterEventSink _eventSink;
}

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
    _eventSink = eventSink;
    return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
    _eventSink = nil;
    return nil;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterFFmpegPlugin* instance = [[FlutterFFmpegPlugin alloc] init];

    FlutterMethodChannel* methodChannel = [FlutterMethodChannel methodChannelWithName:@"flutter_ffmpeg" binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:methodChannel];

    FlutterEventChannel* eventChannel = [FlutterEventChannel eventChannelWithName:@"flutter_ffmpeg_event" binaryMessenger:[registrar messenger]];
    [eventChannel setStreamHandler:instance];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {

    // ARGUMENTS
    NSArray* arguments = call.arguments[@"arguments"];
    NSString* command = call.arguments[@"command"];
    NSString* delimiter = call.arguments[@"delimiter"];

    if ([@"getPlatform" isEqualToString:call.method]) {

        NSString *architecture = [ArchDetect getArch];
        result([FlutterFFmpegPlugin toStringDictionary:KEY_PLATFORM :[NSString stringWithFormat:@"%@-%@", PLATFORM_NAME, architecture]]);

    } else if ([@"getFFmpegVersion" isEqualToString:call.method]) {

        result([FlutterFFmpegPlugin toStringDictionary:KEY_VERSION :[MobileFFmpegConfig getFFmpegVersion]]);

    } else if ([@"executeFFmpegWithArguments" isEqualToString:call.method]) {

        NSLog(@"Running FFmpeg with arguments: %@.\n", arguments);

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{

            int rc = [MobileFFmpeg executeWithArguments:arguments];

            NSLog(@"FFmpeg exited with rc: %d\n", rc);

            result([FlutterFFmpegPlugin toIntDictionary:KEY_RC :[NSNumber numberWithInt:rc]]);
        });

    } else if ([@"executeFFprobeWithArguments" isEqualToString:call.method]) {

        NSLog(@"Running FFprobe with arguments: %@.\n", arguments);

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{

            int rc = [MobileFFprobe executeWithArguments:arguments];

            NSLog(@"FFprobe exited with rc: %d\n", rc);

            result([FlutterFFmpegPlugin toIntDictionary:KEY_RC :[NSNumber numberWithInt:rc]]);
        });

    } else if ([@"cancel" isEqualToString:call.method]) {

        [MobileFFmpeg cancel];

    } else if ([@"enableRedirection" isEqualToString:call.method]) {

        [MobileFFmpegConfig enableRedirection];

    } else if ([@"disableRedirection" isEqualToString:call.method]) {

        [MobileFFmpegConfig disableRedirection];

    } else if ([@"getLogLevel" isEqualToString:call.method]) {

        int logLevel = [MobileFFmpegConfig getLogLevel];
        result([FlutterFFmpegPlugin toIntDictionary:KEY_LOG_LEVEL :[NSNumber numberWithInt:logLevel]]);

    } else if ([@"setLogLevel" isEqualToString:call.method]) {

        NSNumber* logLevel = call.arguments[@"level"];
        [MobileFFmpegConfig setLogLevel:[logLevel intValue]];

    } else if ([@"enableLogs" isEqualToString:call.method]) {

        [MobileFFmpegConfig setLogDelegate:self];

    } else if ([@"disableLogs" isEqualToString:call.method]) {

        [MobileFFmpegConfig setLogDelegate:nil];

    } else if ([@"enableStatistics" isEqualToString:call.method]) {

        [MobileFFmpegConfig setStatisticsDelegate:self];

    } else if ([@"disableStatistics" isEqualToString:call.method]) {

        [MobileFFmpegConfig setStatisticsDelegate:nil];

    } else if ([@"getLastReceivedStatistics" isEqualToString:call.method]) {

        Statistics *statistics = [MobileFFmpegConfig getLastReceivedStatistics];
        result([FlutterFFmpegPlugin toStatisticsDictionary:statistics]);

    } else if ([@"resetStatistics" isEqualToString:call.method]) {

        [MobileFFmpegConfig resetStatistics];

    } else if ([@"setFontconfigConfigurationPath" isEqualToString:call.method]) {

        NSString* path = call.arguments[@"path"];
        [MobileFFmpegConfig setFontconfigConfigurationPath:path];

    } else if ([@"setFontDirectory" isEqualToString:call.method]) {

        NSString* fontDirectoryPath = call.arguments[@"fontDirectory"];
        NSDictionary* fontNameMapping = call.arguments[@"fontNameMap"];
        [MobileFFmpegConfig setFontDirectory:fontDirectoryPath with:fontNameMapping];

    } else if ([@"getPackageName" isEqualToString:call.method]) {

        NSString *packageName = [MobileFFmpegConfig getPackageName];
        result([FlutterFFmpegPlugin toStringDictionary:KEY_PACKAGE_NAME :packageName]);

    } else if ([@"getExternalLibraries" isEqualToString:call.method]) {

        NSArray *externalLibraries = [MobileFFmpegConfig getExternalLibraries];
        result(externalLibraries);

    } else if ([@"getLastReturnCode" isEqualToString:call.method]) {

        int lastReturnCode = [MobileFFmpegConfig getLastReturnCode];
        result([FlutterFFmpegPlugin toIntDictionary:KEY_LAST_RC :[NSNumber numberWithInt:lastReturnCode]]);

    } else if ([@"getLastCommandOutput" isEqualToString:call.method]) {

        NSString *lastCommandOutput = [MobileFFmpegConfig getLastCommandOutput];
        result([FlutterFFmpegPlugin toStringDictionary:KEY_LAST_COMMAND_OUTPUT :lastCommandOutput]);

    } else if ([@"getMediaInformation" isEqualToString:call.method]) {

        NSString* path = call.arguments[@"path"];

        NSLog(@"Getting media information for %@.\n", path);

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            MediaInformation *mediaInformation = [MobileFFprobe getMediaInformation:path];
            result([FlutterFFmpegPlugin toMediaInformationDictionary:mediaInformation]);
        });

    } else if ([@"registerNewFFmpegPipe" isEqualToString:call.method]) {

        NSString *pipe = [MobileFFmpegConfig registerNewFFmpegPipe];
        result([FlutterFFmpegPlugin toStringDictionary:KEY_PIPE :pipe]);

    } else {

        result(FlutterMethodNotImplemented);

    }
}

- (void)logCallback: (int)level :(NSString*)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
        dictionary[KEY_LOG_LEVEL] = [NSNumber numberWithInt:level];
        dictionary[KEY_LOG_TEXT] = message;

        [self emitLogMessage: dictionary];
    });
}

- (void)statisticsCallback:(Statistics *)statistics {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self emitStatistics: statistics];
    });
}

- (void)emitLogMessage:(NSDictionary*)logMessage{
    _eventSink([FlutterFFmpegPlugin toStringDictionary:EVENT_LOG withDictionary:logMessage]);
}

- (void)emitStatistics:(Statistics*)statistics{
    NSDictionary *dictionary = [FlutterFFmpegPlugin toStatisticsDictionary:statistics];
    _eventSink([FlutterFFmpegPlugin toStringDictionary:EVENT_STAT withDictionary:dictionary]);
}

+ (NSDictionary *)toStringDictionary:(NSString*)key :(NSString*)value {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    dictionary[key] = value;

    return dictionary;
}

+ (NSDictionary *)toStringDictionary:(NSString*)key withDictionary:(NSDictionary*)value {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    dictionary[key] = value;

    return dictionary;
}

+ (NSDictionary *)toIntDictionary:(NSString*)key :(NSNumber*)value {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    dictionary[key] = value;

    return dictionary;
}

+ (NSDictionary *)toStatisticsDictionary:(Statistics*)statistics {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];

    if (statistics != nil) {
        dictionary[KEY_STAT_TIME] = [NSNumber numberWithInt: [statistics getTime]];
        dictionary[KEY_STAT_SIZE] = [NSNumber numberWithLong: [statistics getSize]];

        dictionary[KEY_STAT_BITRATE] = [NSNumber numberWithDouble: [statistics getBitrate]];
        dictionary[KEY_STAT_SPEED] = [NSNumber numberWithDouble: [statistics getSpeed]];

        dictionary[KEY_STAT_VIDEO_FRAME_NUMBER] = [NSNumber numberWithInt: [statistics getVideoFrameNumber]];
        dictionary[KEY_STAT_VIDEO_QUALITY] = [NSNumber numberWithFloat: [statistics getVideoQuality]];
        dictionary[KEY_STAT_VIDEO_FPS] = [NSNumber numberWithFloat: [statistics getVideoFps]];
    }

    return dictionary;
}

+ (NSDictionary *)toMediaInformationDictionary:(MediaInformation*)mediaInformation {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];

    if (mediaInformation != nil) {
        if ([mediaInformation getFormat] != nil) {
            dictionary[@"format"] =  [mediaInformation getFormat];
        }
        if ([mediaInformation getPath] != nil) {
            dictionary[@"path"] = [mediaInformation getPath];
        }
        if ([mediaInformation getStartTime] != nil) {
            dictionary[@"startTime"] = [mediaInformation getStartTime];
        }
        if ([mediaInformation getDuration] != nil) {
            dictionary[@"duration"] = [mediaInformation getDuration];
        }
        if ([mediaInformation getBitrate] != nil) {
            dictionary[@"bitrate"] = [mediaInformation getBitrate];
        }
        if ([mediaInformation getRawInformation] != nil) {
            dictionary[@"rawInformation"] = [mediaInformation getRawInformation];
        }

        NSDictionary *metadataDictionary = [mediaInformation getMetadataEntries];
        if (metadataDictionary != nil && ([metadataDictionary count] > 0)) {
            dictionary[@"metadata"] = metadataDictionary;
        }

        NSArray *streams = [mediaInformation getStreams];
        if (streams != nil && ([streams count] > 0)) {
            NSMutableArray *array = [[NSMutableArray alloc] init];

            for (int i=0; i < [streams count]; i++) {
                StreamInformation *streamInformation= [streams objectAtIndex:i];
                [array addObject: [FlutterFFmpegPlugin toStreamInformationDictionary:streamInformation]];
            }

            dictionary[@"streams"] = array;
        }
    }

    return dictionary;
}

+ (NSDictionary *)toStreamInformationDictionary:(StreamInformation*)streamInformation {
    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];

    if (streamInformation != nil) {
        if ([streamInformation getIndex] != nil) {
            dictionary[@"index"] = [streamInformation getIndex];
        }
        if ([streamInformation getType] != nil) {
            dictionary[@"type"] = [streamInformation getType];
        }
        if ([streamInformation getCodec] != nil) {
            dictionary[@"codec"] = [streamInformation getCodec];
        }
        if ([streamInformation getFullCodec] != nil) {
            dictionary[@"fullCodec"] = [streamInformation getFullCodec];
        }
        if ([streamInformation getFormat] != nil) {
            dictionary[@"format"] = [streamInformation getFormat];
        }
        if ([streamInformation getFullFormat] != nil) {
            dictionary[@"fullFormat"] = [streamInformation getFullFormat];
        }
        if ([streamInformation getWidth] != nil) {
            dictionary[@"width"] = [streamInformation getWidth];
        }
        if ([streamInformation getHeight] != nil) {
            dictionary[@"height"] = [streamInformation getHeight];
        }
        if ([streamInformation getBitrate] != nil) {
            dictionary[@"bitrate"] = [streamInformation getBitrate];
        }
        if ([streamInformation getSampleRate] != nil) {
            dictionary[@"sampleRate"] = [streamInformation getSampleRate];
        }
        if ([streamInformation getSampleFormat] != nil) {
            dictionary[@"sampleFormat"] = [streamInformation getSampleFormat];
        }
        if ([streamInformation getChannelLayout] != nil) {
            dictionary[@"channelLayout"] = [streamInformation getChannelLayout];
        }
        if ([streamInformation getSampleAspectRatio] != nil) {
            dictionary[@"sampleAspectRatio"] = [streamInformation getSampleAspectRatio];
        }
        if ([streamInformation getDisplayAspectRatio] != nil) {
            dictionary[@"displayAspectRatio"] = [streamInformation getDisplayAspectRatio];
        }
        if ([streamInformation getAverageFrameRate] != nil) {
            dictionary[@"averageFrameRate"] = [streamInformation getAverageFrameRate];
        }
        if ([streamInformation getRealFrameRate] != nil) {
            dictionary[@"realFrameRate"] = [streamInformation getRealFrameRate];
        }
        if ([streamInformation getTimeBase] != nil) {
            dictionary[@"timeBase"] = [streamInformation getTimeBase];
        }
        if ([streamInformation getCodecTimeBase] != nil) {
            dictionary[@"codecTimeBase"] = [streamInformation getCodecTimeBase];
        }

        NSDictionary *metadataDictionary = [streamInformation getMetadataEntries];
        if (metadataDictionary != nil && ([metadataDictionary count] > 0)) {
            dictionary[@"metadata"] = metadataDictionary;
        }

        NSDictionary *sidedataDictionary = [streamInformation getSidedataEntries];
        if (sidedataDictionary != nil && ([sidedataDictionary count] > 0)) {
            dictionary[@"sidedata"] = sidedataDictionary;
        }
    }

    return dictionary;
}

@end

File: flutter_sound/android/src/main/java/com/dooboolab/ffmpeg/FlutterFFmpegPlugin.java
/*
 * Copyright (c) 2019 Taner Sener
 *
 * This file is part of FlutterFFmpeg.
 *
 * FlutterFFmpeg is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FlutterFFmpeg is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with FlutterFFmpeg.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.dooboolab.ffmpeg;

import android.content.Context;

import com.arthenica.mobileffmpeg.AbiDetect;
import com.arthenica.mobileffmpeg.Config;
import com.arthenica.mobileffmpeg.FFmpeg;
import com.arthenica.mobileffmpeg.Level;
import com.arthenica.mobileffmpeg.LogCallback;
import com.arthenica.mobileffmpeg.LogMessage;
import com.arthenica.mobileffmpeg.MediaInformation;
import com.arthenica.mobileffmpeg.Statistics;
import com.arthenica.mobileffmpeg.StatisticsCallback;
import com.arthenica.mobileffmpeg.StreamInformation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * <h3>Flutter FFmpeg Plugin</h3>
 *
 * @author Taner Sener
 * @since 0.1.0
 */
public class FlutterFFmpegPlugin implements MethodCallHandler, EventChannel.StreamHandler {
    public static final String LIBRARY_NAME = "flutter-ffmpeg";

    public static final String PLATFORM_NAME = "android";
    public static final String KEY_VERSION = "version";
    public static final String KEY_RC = "rc";
    public static final String KEY_PLATFORM = "platform";
    public static final String KEY_PACKAGE_NAME = "packageName";
    public static final String KEY_LAST_RC = "lastRc";
    public static final String KEY_PIPE = "pipe";

    public static final String KEY_LAST_COMMAND_OUTPUT = "lastCommandOutput";
    public static final String KEY_LOG_TEXT = "log";

    public static final String KEY_LOG_LEVEL = "level";
    public static final String KEY_STAT_TIME = "time";
    public static final String KEY_STAT_SIZE = "size";
    public static final String KEY_STAT_BITRATE = "bitrate";
    public static final String KEY_STAT_SPEED = "speed";
    public static final String KEY_STAT_VIDEO_FRAME_NUMBER = "videoFrameNumber";
    public static final String KEY_STAT_VIDEO_QUALITY = "videoQuality";
    public static final String KEY_STAT_VIDEO_FPS = "videoFps";

    public static final String EVENT_LOG = "FlutterFFmpegLogCallback";
    public static final String EVENT_STAT = "FlutterFFmpegStatisticsCallback";

    private EventChannel.EventSink eventSink;
    //private final Registrar registrar;
    private final FlutterFFmpegResultHandler flutterFFmpegResultHandler;

    static FlutterFFmpegPlugin flutterFFmpegPlugin; // singleton
    static Context              androidContext;


    public static void attachFFmpegPlugin( Context ctx, BinaryMessenger messenger )
    {
        assert ( flutterFFmpegPlugin == null );
        flutterFFmpegPlugin = new FlutterFFmpegPlugin();
        final MethodChannel channel = new MethodChannel(messenger, "flutter_ffmpeg");
        channel.setMethodCallHandler(flutterFFmpegPlugin);
        androidContext = ctx;
        final EventChannel eventChannel = new EventChannel(messenger, "flutter_ffmpeg_event");
        eventChannel.setStreamHandler(flutterFFmpegPlugin);
    }



    /**
     * Registers plugin to registry.
     */
    /*
    public static void registerWith(final Registrar registrar) {
        FlutterFFmpegPlugin flutterFFmpegPlugin = new FlutterFFmpegPlugin(registrar);

        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_ffmpeg");
        channel.setMethodCallHandler(flutterFFmpegPlugin);

        final EventChannel eventChannel = new EventChannel(registrar.messenger(), "flutter_ffmpeg_event");
        eventChannel.setStreamHandler(flutterFFmpegPlugin);
    }
*/
    private FlutterFFmpegPlugin() {


        this.flutterFFmpegResultHandler = new FlutterFFmpegResultHandler();
    }


    private Context getActiveContext() {
        return androidContext;
    }

    /**
     * Handles method calls.
     *
     * @param call   method call
     * @param result result callback
     */
    @Override
    public void onMethodCall(final MethodCall call, final Result result) {
        if (call.method.equals("getPlatform")) {

            final String abi = AbiDetect.getAbi();
            flutterFFmpegResultHandler.success(result, toStringMap(KEY_PLATFORM, PLATFORM_NAME + "-" + abi));

        } else if (call.method.equals("getFFmpegVersion")) {

            final String version = Config.getFFmpegVersion();
            flutterFFmpegResultHandler.success(result, toStringMap(KEY_VERSION, version));

        } else if (call.method.equals("executeFFmpegWithArguments")) {

            List<String> arguments = call.argument("arguments");

            final FlutterFFmpegExecuteFFmpegAsyncArgumentsTask asyncTask = new FlutterFFmpegExecuteFFmpegAsyncArgumentsTask(arguments, flutterFFmpegResultHandler, result);
            asyncTask.execute("dummy-trigger");

        } else if (call.method.equals("executeFFprobeWithArguments")) {

            List<String> arguments = call.argument("arguments");

            final FlutterFFmpegExecuteFFprobeAsyncArgumentsTask asyncTask = new FlutterFFmpegExecuteFFprobeAsyncArgumentsTask(arguments, flutterFFmpegResultHandler, result);
            asyncTask.execute("dummy-trigger");

        } else if (call.method.equals("cancel")) {

            FFmpeg.cancel();

        } else if (call.method.equals("enableRedirection")) {

            Config.enableRedirection();

        } else if (call.method.equals("disableRedirection")) {

            Config.disableRedirection();

        } else if (call.method.equals("getLogLevel")) {

            final Level level = Config.getLogLevel();
            flutterFFmpegResultHandler.success(result, toIntMap(KEY_LOG_LEVEL, levelToInt(level)));

        } else if (call.method.equals("setLogLevel")) {

            Integer level = call.argument("level");
            if (level == null) {
                level = Level.AV_LOG_TRACE.getValue();
            }
            Config.setLogLevel(Level.from(level));

        } else if (call.method.equals("enableLogs")) {

            Config.enableLogCallback(new LogCallback() {

                @Override
                public void apply(final LogMessage logMessage) {
                    emitLogMessage(logMessage);
                }
            });

        } else if (call.method.equals("disableLogs")) {

            Config.enableLogCallback(null);

        } else if (call.method.equals("enableStatistics")) {

            Config.enableStatisticsCallback(new StatisticsCallback() {

                @Override
                public void apply(final Statistics statistics) {
                    emitStatistics(statistics);
                }
            });

        } else if (call.method.equals("disableStatistics")) {

            Config.enableStatisticsCallback(null);

        } else if (call.method.equals("getLastReceivedStatistics")) {

            flutterFFmpegResultHandler.success(result, toMap(Config.getLastReceivedStatistics()));

        } else if (call.method.equals("resetStatistics")) {

            Config.resetStatistics();

        } else if (call.method.equals("setFontconfigConfigurationPath")) {
            String path = call.argument("path");

            Config.setFontconfigConfigurationPath(path);

        } else if (call.method.equals("setFontDirectory")) {

            String path = call.argument("fontDirectory");
            Map<String, String> map = call.argument("fontNameMap");

            Config.setFontDirectory(getActiveContext(), path, map);

        } else if (call.method.equals("getPackageName")) {

            final String packageName = Config.getPackageName();
            flutterFFmpegResultHandler.success(result, toStringMap(KEY_PACKAGE_NAME, packageName));

        } else if (call.method.equals("getExternalLibraries")) {

            final List<String> externalLibraries = Config.getExternalLibraries();
            flutterFFmpegResultHandler.success(result, externalLibraries);

        } else if (call.method.equals("getLastReturnCode")) {

            int lastReturnCode = Config.getLastReturnCode();
            flutterFFmpegResultHandler.success(result, toIntMap(KEY_LAST_RC, lastReturnCode));

        } else if (call.method.equals("getLastCommandOutput")) {

            final String lastCommandOutput = Config.getLastCommandOutput();
            flutterFFmpegResultHandler.success(result, toStringMap(KEY_LAST_COMMAND_OUTPUT, lastCommandOutput));

        } else if (call.method.equals("getMediaInformation")) {
            final String path = call.argument("path");
            Integer timeout = call.argument("timeout");
            if (timeout == null) {
                timeout = 10000;
            }

            final FlutterFFmpegGetMediaInformationAsyncTask asyncTask = new FlutterFFmpegGetMediaInformationAsyncTask(path, flutterFFmpegResultHandler, result);
            asyncTask.execute();

        } else if (call.method.equals("registerNewFFmpegPipe")) {

            final String pipe = Config.registerNewFFmpegPipe(getActiveContext());
            flutterFFmpegResultHandler.success(result, toStringMap(KEY_PIPE, pipe));

        } else {
            flutterFFmpegResultHandler.notImplemented(result);
        }
    }

    @Override
    public void onListen(Object o, EventChannel.EventSink eventSink) {
        this.eventSink = eventSink;
    }

    @Override
    public void onCancel(Object o) {
        this.eventSink = null;
    }

    protected void emitLogMessage(final LogMessage logMessage) {
        final HashMap<String, Object> logWrapperMap = new HashMap<>();
        final HashMap<String, Object> logMap = new HashMap<>();

        logMap.put(KEY_LOG_LEVEL, levelToInt(logMessage.getLevel()));
        logMap.put(KEY_LOG_TEXT, logMessage.getText());

        logWrapperMap.put(EVENT_LOG, logMap);

        flutterFFmpegResultHandler.success(eventSink, logWrapperMap);
    }

    protected void emitStatistics(final Statistics statistics) {
        final HashMap<String, Object> statisticsMap = new HashMap<>();
        statisticsMap.put(EVENT_STAT, toMap(statistics));
        flutterFFmpegResultHandler.success(eventSink, statisticsMap);
    }

    public static int levelToInt(final Level level) {
        return (level == null) ? Level.AV_LOG_TRACE.getValue() : level.getValue();
    }

    public static HashMap<String, String> toStringMap(final String key, final String value) {
        final HashMap<String, String> map = new HashMap<>();
        map.put(key, value);
        return map;
    }

    public static HashMap<String, Integer> toIntMap(final String key, final int value) {
        final HashMap<String, Integer> map = new HashMap<>();
        map.put(key, value);
        return map;
    }

    public static Map<String, Object> toMap(final Statistics statistics) {
        final HashMap<String, Object> statisticsMap = new HashMap<>();

        if (statistics != null) {
            statisticsMap.put(KEY_STAT_TIME, statistics.getTime());
            statisticsMap.put(KEY_STAT_SIZE, (statistics.getSize() < Integer.MAX_VALUE) ? (int) statistics.getSize() : (int) (statistics.getSize() % Integer.MAX_VALUE));
            statisticsMap.put(KEY_STAT_BITRATE, statistics.getBitrate());
            statisticsMap.put(KEY_STAT_SPEED, statistics.getSpeed());

            statisticsMap.put(KEY_STAT_VIDEO_FRAME_NUMBER, statistics.getVideoFrameNumber());
            statisticsMap.put(KEY_STAT_VIDEO_QUALITY, statistics.getVideoQuality());
            statisticsMap.put(KEY_STAT_VIDEO_FPS, statistics.getVideoFps());
        }

        return statisticsMap;
    }

    public static HashMap<String, Object> toMediaInformationMap(final MediaInformation mediaInformation) {
        final HashMap<String, Object> map = new HashMap<>();

        if (mediaInformation != null) {
            if (mediaInformation.getFormat() != null) {
                map.put("format", mediaInformation.getFormat());
            }
            if (mediaInformation.getPath() != null) {
                map.put("path", mediaInformation.getPath());
            }
            if (mediaInformation.getStartTime() != null) {
                map.put("startTime", mediaInformation.getStartTime().intValue());
            }
            if (mediaInformation.getDuration() != null) {
                map.put("duration", mediaInformation.getDuration().intValue());
            }
            if (mediaInformation.getBitrate() != null) {
                map.put("bitrate", mediaInformation.getBitrate().intValue());
            }
            if (mediaInformation.getRawInformation() != null) {
                map.put("rawInformation", mediaInformation.getRawInformation());
            }

            final Set<Map.Entry<String, String>> metadata = mediaInformation.getMetadataEntries();
            if ((metadata != null) && (metadata.size() > 0)) {
                final HashMap<String, String> metadataMap = new HashMap<>();

                for (Map.Entry<String, String> entry : metadata) {
                    metadataMap.put(entry.getKey(), entry.getValue());
                }

                map.put("metadata", metadataMap);
            }

            final List<StreamInformation> streams = mediaInformation.getStreams();
            if ((streams != null) && (streams.size() > 0)) {
                final ArrayList<Map<String, Object>> array = new ArrayList<>();

                for (StreamInformation streamInformation : streams) {
                    array.add(toStreamInformationMap(streamInformation));
                }

                map.put("streams", array);
            }
        }

        return map;
    }

    public static Map<String, Object> toStreamInformationMap(final StreamInformation streamInformation) {
        final HashMap<String, Object> map = new HashMap<>();

        if (streamInformation != null) {
            if (streamInformation.getIndex() != null) {
                map.put("index", streamInformation.getIndex().intValue());
            }
            if (streamInformation.getType() != null) {
                map.put("type", streamInformation.getType());
            }
            if (streamInformation.getCodec() != null) {
                map.put("codec", streamInformation.getCodec());
            }
            if (streamInformation.getFullCodec() != null) {
                map.put("fullCodec", streamInformation.getFullCodec());
            }
            if (streamInformation.getFormat() != null) {
                map.put("format", streamInformation.getFormat());
            }
            if (streamInformation.getFullFormat() != null) {
                map.put("fullFormat", streamInformation.getFullFormat());
            }
            if (streamInformation.getWidth() != null) {
                map.put("width", streamInformation.getWidth().intValue());
            }
            if (streamInformation.getHeight() != null) {
                map.put("height", streamInformation.getHeight().intValue());
            }
            if (streamInformation.getBitrate() != null) {
                map.put("bitrate", streamInformation.getBitrate().intValue());
            }
            if (streamInformation.getSampleRate() != null) {
                map.put("sampleRate", streamInformation.getSampleRate().intValue());
            }
            if (streamInformation.getSampleFormat() != null) {
                map.put("sampleFormat", streamInformation.getSampleFormat());
            }
            if (streamInformation.getChannelLayout() != null) {
                map.put("channelLayout", streamInformation.getChannelLayout());
            }
            if (streamInformation.getSampleAspectRatio() != null) {
                map.put("sampleAspectRatio", streamInformation.getSampleAspectRatio());
            }
            if (streamInformation.getDisplayAspectRatio() != null) {
                map.put("displayAspectRatio", streamInformation.getDisplayAspectRatio());
            }
            if (streamInformation.getAverageFrameRate() != null) {
                map.put("averageFrameRate", streamInformation.getAverageFrameRate());
            }
            if (streamInformation.getRealFrameRate() != null) {
                map.put("realFrameRate", streamInformation.getRealFrameRate());
            }
            if (streamInformation.getTimeBase() != null) {
                map.put("timeBase", streamInformation.getTimeBase());
            }
            if (streamInformation.getCodecTimeBase() != null) {
                map.put("codecTimeBase", streamInformation.getCodecTimeBase());
            }

            final Set<Map.Entry<String, String>> metadata = streamInformation.getMetadataEntries();
            if ((metadata != null) && (metadata.size() > 0)) {
                final HashMap<String, String> metadataMap = new HashMap<>();

                for (Map.Entry<String, String> entry : metadata) {
                    metadataMap.put(entry.getKey(), entry.getValue());
                }

                map.put("metadata", metadataMap);
            }

            final Set<Map.Entry<String, String>> sidedata = streamInformation.getSidedataEntries();
            if ((sidedata != null) && (sidedata.size() > 0)) {
                final HashMap<String, String> sidedataMap = new HashMap<>();

                for (Map.Entry<String, String> entry : sidedata) {
                    sidedataMap.put(entry.getKey(), entry.getValue());
                }

                map.put("sidedata", sidedataMap);
            }
        }

        return map;
    }

}
